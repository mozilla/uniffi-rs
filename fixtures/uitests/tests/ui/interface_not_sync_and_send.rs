
use std::cell::Cell;

// Unfortunately, path is relative to a temporary build directory :-/
uniffi_macros::generate_and_include_scaffolding!("../../../fixtures/uitests/src/counter.udl");

fn main() { /* empty main required by `trybuild` */}

pub struct Counter {
    // This will fail to compile, because `Cell` is not `Sync`.
    value: Cell<u32>,
}

impl Counter {
    pub fn new() -> Self {
        Self { value: Cell::new(0) }
    }

    pub fn increment(&self) -> u32 {
        let new_value = self.value.get() + 1;
        self.value.set(new_value);
        new_value
    }
}
