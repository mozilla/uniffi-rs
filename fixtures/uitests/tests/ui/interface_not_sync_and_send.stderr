error[E0277]: `Cell<u32>` cannot be shared between threads safely
   --> $DIR/counter.uniffi.rs:119:1
    |
119 | uniffi::deps::static_assertions::assert_impl_all!(Counter: Sync, Send);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | |
    | `Cell<u32>` cannot be shared between threads safely
    | required by this bound in `assert_impl_all`
    |
    = help: within `Counter`, the trait `Sync` is not implemented for `Cell<u32>`
    = note: required because it appears within the type `Counter`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Arc<Counter>: FfiConverter` is not satisfied
   --> $DIR/counter.uniffi.rs:145:9
    |
145 |         <std::sync::Arc<Counter> as uniffi::FfiConverter>::lower(_arc)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `FfiConverter` is not implemented for `Arc<Counter>`
    |
    = help: the following implementations were found:
              <Arc<T> as FfiConverter>
    = note: required by `lower`

error[E0277]: the trait bound `Arc<Counter>: FfiConverter` is not satisfied
   --> $DIR/counter.uniffi.rs:160:14
    |
160 |             &<std::sync::Arc<Counter> as uniffi::FfiConverter>::try_lift(ptr).unwrap(),
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `FfiConverter` is not implemented for `Arc<Counter>`
    |
    = help: the following implementations were found:
              <Arc<T> as FfiConverter>
    = note: required by `try_lift`
